distributed/tests/test_asyncprocess.py

```diff
@@ -297,3 +299,124 @@ def test_terminate_after_stop():
+
+    with pristine_loop() as loop:
+        try:
+            loop.run_sync(gen.coroutine(parent_process_coroutine), timeout=10)
+        finally:
+            loop.stop()
+
+            raise RuntimeError("this should be unreachable due to os._exit")
+
+
+def test_asyncprocess_child_teardown_on_parent_exit():
+    """ Check that a child process started by AsyncProcess exits if its parent
+    exits.
+
+    The motivation is to ensure that if an AsyncProcess is created and the
+    creator process dies unexpectedly (e.g, via Out-of-memory SIGKILL), the
+    child process and resources held by it should not be leaked.
+
+    The child should monitor its parent and exit promptly if the parent exits.
+
+    [test process] -> [parent using AsyncProcess (dies)] -> [worker process]
+                 \                                          /
+                  \________ <--   child_pipe   <-- ________/
+    """
+    # When child_pipe is closed, the children_alive pipe unblocks.
+    children_alive, child_pipe = mp_context.Pipe(duplex=False)
+
+    try:
+        parent = mp_context.Process(target=_parent_process, args=(child_pipe,))
+        parent.start()
+
+        # Close our reference to child_pipe so that the child has the only one.
+        child_pipe.close()
+
+        # Wait for the parent to exit. By the time join returns, the child
+        # process is orphaned, and should be in the process of exiting by
+        # itself.
+        parent.join()
+
+        # By the time we reach here,the parent has exited. The parent only exits
+        # when the child is ready to enter the sleep, so all of the slow things
+        # (process startup, etc) should have happened by now, even on a busy
+        # system. A short timeout should therefore be appropriate.
+        short_timeout = 5.
+        # Poll is used to allow other tests to proceed after this one in case of
+        # test failure.
+        try:
+            readable = children_alive.poll(short_timeout)
+        except EnvironmentError:
+            # Windows can raise BrokenPipeError. EnvironmentError is caught for
+            # Python2/3 portability.
+            assert sys.platform.startswith('win'), "should not raise on windows"
+            # Broken pipe implies closed, which is readable.
+            readable = True
+
+        # If this assert fires, then something went wrong. Either the child
+        # should write into the pipe, or it should exit and the pipe should be
+        # closed (which makes it become readable).
+        assert readable
+
+        try:
+            # This won't block due to the above 'assert readable'.
+            result = children_alive.recv()
+        except EOFError:
+            pass # Test passes.
+        except EnvironmentError:
+            # Windows can raise BrokenPipeError. EnvironmentError is caught for
+            # Python2/3 portability.
+            assert sys.platform.startswith('win'), "should not raise on windows"
```

Comment:
"Should not raise on Windows" should probably be "Should not raise on non-Windows".